edgexfoundry/edgex-go
# feat(security): Implement spiffe token provider services

- add spire server Dockerfile
- add spire agent Dockerfile
- add spire config Dockerfile
- add spiffe token provider and `gettoken` restful endpoint


Closes: #3773

Signed-off-by: Jim Wang <yutsung.jim.wang@intel.com>







**If your build fails**  due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x ] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [ ] I am not introducing a new dependency (add notes below if you are)    _spiffe/v2 and its related dependencies_ 
- [ ] I have added unit tests for the new feature or bug fix (if not, why?)   x509Source from spiffe APIs not testable directly as they don't have interface to mock on
  <link to docs PR>

## Testing Instructions

With updated docker-compose files from this branch (https://github.com/jim-wang-intel/edgex-compose/tree/add-spiffe-samples) and do `make run dev delayed-start` from the command line , then follow the testcase.txt under security-spiffe-token-provider/ service: https://github.com/jim-wang-intel/edgex-go/blob/add-spiffe-token-provider/cmd/security-spiffe-token-provider/testcase.txt 
to verify you can get a vault token.

## New Dependency Instructions (If applicable)

New dependency
- go-spiffe/v2
```
jim@yutsungj-DESK2:~/go/src/github.com/edgexfoundry/edgex-go$ GO111MODULE=on go list -m all | grep spiffe
github.com/spiffe/go-spiffe/v2 v2.0.0-beta.11
```

Total increase in new imports: 8
    Does the new import introduce additional import dependencies, if so, how many?
       Yes, totally **7** in addition to `go-spiffe/v2`
```
GO111MODULE=on go mod graph | grep spiffe
github.com/edgexfoundry/edgex-go github.com/spiffe/go-spiffe/v2@v2.0.0-beta.11
github.com/spiffe/go-spiffe/v2@v2.0.0-beta.11 github.com/golang/protobuf@v1.4.2
github.com/spiffe/go-spiffe/v2@v2.0.0-beta.11 github.com/stretchr/testify@v1.5.1
github.com/spiffe/go-spiffe/v2@v2.0.0-beta.11 github.com/zeebo/errs@v1.2.2
github.com/spiffe/go-spiffe/v2@v2.0.0-beta.11 google.golang.org/grpc@v1.33.2
github.com/spiffe/go-spiffe/v2@v2.0.0-beta.11 google.golang.org/grpc/examples@v0.0.0-20201130180447-c456688b1860
github.com/spiffe/go-spiffe/v2@v2.0.0-beta.11 google.golang.org/protobuf@v1.25.0
github.com/spiffe/go-spiffe/v2@v2.0.0-beta.11 gopkg.in/square/go-jose.v2@v2.4.1
```
    Releases/Tags: 19
    Contributors: 22
    License - what is the license, and is it Apache 2.0 compatible?  Apache License see https://github.com/spiffe/go-spiffe/blob/main/LICENSE 
    Stars/Forks/Watchers: 71 stars / 43 Forks / 20 Watchers
```
 godocs: https://pkg.go.dev/github.com/spiffe/go-spiffe/v2
